lib/internal/http2/core.js

```diff
@@ -1978,19 +1978,55 @@ class Http2Stream extends Duplex {
 
     let req;
 
-    // writeGeneric does not destroy on error and we cannot enable autoDestroy,
-    // so make sure to destroy on error.
-    const callback = (err) => {
+    let waitingForWriteCallback = true;
+    let waitingForEndCheck = true;
+    let writeCallbackErr;
+    let endCheckCallbackErr;
+    const done = () => {
+      if (waitingForEndCheck || waitingForWriteCallback) return;
+      const err = writeCallbackErr || endCheckCallbackErr;
+      // writeGeneric does not destroy on error and
+      // we cannot enable autoDestroy,
+      // so make sure to destroy on error.
       if (err) {
         this.destroy(err);
       }
       cb(err);
     };
+    const writeCallback = (err) => {
+      waitingForWriteCallback = false;
+      writeCallbackErr = err;
+      done();
+    };
+    const endCheckCallback = (err) => {
+      waitingForEndCheck = false;
+      endCheckCallbackErr = err;
+      done();
+    };
+    // Shutdown write stream right after last chunk is sent
+    // so final DATA frame can include END_STREAM flag
+    process.nextTick(() => {
+      if (writeCallbackErr ||
+        !this._writableState.ending ||
+        this._writableState.buffered.length ||
+        (this[kState].flags & STREAM_FLAGS_HAS_TRAILERS))
+        return endCheckCallback();
+      const handle = this[kHandle];
+      if (!handle) return endCheckCallback();
+      debugStreamObj(this, 'kWriteGeneric shutting down');
+      const req = new ShutdownWrap();
+      req.oncomplete = afterShutdown;
+      req.callback = endCheckCallback;
+      req.handle = handle;
+      const err = handle.shutdown(req);
```

Comment:
Maybe I’m missing something, but won’t this conflict with the existing `handle.shutdown()` call from `_final`?

Reply:
I think you're right. It my testing it didn't happen because the `handle` was `undefined` by the time it reached `_final`, meaning the object is destroyed before `final` gets emitted. But that be because no test continues to read after ending the `Writable` stream. I'll test to see if I can reproduce that, but add a `shutdownCalled` flag regardless.

Reply:
It calls it twice, but there's no real conflict because `http2.cc` guards against multiple shutdown calls. But it would still be cleaner to set a flag to avoid it calling it twice.

Reply:
Right … it’s still not really within the API contract of `StreamBase` to call `.shutdown()` while a write is still active – but as you pointed out, it’s not really trivial to do something like zlib does here :/

Reply:
Without resorting to using `::DoShutDown()`, we can also add an `isFinal` variable to `WriteWrap`. We can either modify the `StreamBase` API to include the argument in `_writev()` and `_write()`, modify `writevGeneric()` and `writeGeneric()`, or fork a new `stream_base_commons.js` file just for HTTP2.